### PR TITLE
fix/adata: fix pushing duplicate entries in AppendOnly Data

### PR DESCRIPTION
**Pull request by Yoga07**
Resolves maidsafe/safe_client_libs#875

### CODE REVIEW

File: src/append_only_data.rs [hunk 1]
```diff
@@ -676,6 +698,28 @@ where
     P: Permissions + std::hash::Hash + Clone,
 {
     fn append(&mut self, entries: &[(Vec<u8>, Vec<u8>)]) -> Result<()> {
+        let existing: BTreeSet<Vec<u8>> = self
```
Comment (lionel-faber) [inline src/append_only_data.rs]: This block seems to be the same as the previous one.
We can consider putting it into a function.

File: src/errors.rs [hunk 2]
```diff
@@ -40,6 +41,10 @@ pub enum Error {
     InvalidEntryActions(BTreeMap<Vec<u8>, EntryError>),
     /// Key does not exist
     NoSuchKey,
+    /// Key already exists
+    KeyExists(Entries),
```
Comment (lionel-faber) [inline src/errors.rs]: I think this is better named as `KeysExist`

File: src/errors.rs [hunk 3]
```diff
@@ -40,6 +41,10 @@ pub enum Error {
...
+    /// Duplicate Entries in this push
+    DuplicateEntries,
```
Comment (lionel-faber) [inline src/errors.rs]: This error is returned if the _keys_ are duplicate.

File: src/append_only_data.rs [hunk 4]
```diff
@@ -663,6 +663,28 @@ where
     P: Permissions + std::hash::Hash + Clone,
 {
     fn append(&mut self, entries: &[(Vec<u8>, Vec<u8>)], last_entries_index: u64) -> Result<()> {
+        let existing: BTreeSet<Vec<u8>> = self
```
Comment (Fraser999) [inline src/append_only_data.rs]: `existing` isn't used until line 679 below, so better to create it just immediately before there rather than this early.  (e.g. this would would be wasted effort if we return at line 676).

File: src/append_only_data.rs [hunk 5]
```diff
@@ -663,6 +663,28 @@ where
...
+            .inner
+            .data
+            .iter()
+            .map(|tuple| tuple.0.clone())
```
Comment (Fraser999) [inline src/append_only_data.rs]: I think we should be able to avoid cloning the keys of the entries if we make the type of `existing` and `new` `BTreeSet<&Vec<u8>>`.

Also, a minor point: to make the code potentially more readable, you could change the captured variable from `tuple` to `(key, _value)`, then rather than calling `tuple.0` you'd just have `key`.  Not a big deal for something as simple as this scenario though.

File: src/append_only_data.rs [hunk 6]
```diff
@@ -663,6 +663,28 @@ where
     P: Permissions + std::hash::Hash + Clone,
 {
     fn append(&mut self, entries: &[(Vec<u8>, Vec<u8>)], last_entries_index: u64) -> Result<()> {
```
Comment (Fraser999) [inline src/append_only_data.rs]: I think it may be better to just pass a `Vec` here rather than a slice, since the collection will be consumed by this function.  I.e. changing the signature to `mut entries: Entries`.

File: src/append_only_data.rs [hunk 7]
```diff
@@ -663,6 +663,28 @@ where
...
+            .collect();
+        let new: BTreeSet<Vec<u8>> = entries.iter().map(|tuple| tuple.0.clone()).collect();
+
+        // If duplicate entries are present in the push.
+        if new.len() < entries.len() {
+            return Err(Error::DuplicateEntries);
+        }
+
+        if !existing.is_disjoint(&new) {
+            let dup: Entries = entries
```
Comment (Fraser999) [inline src/append_only_data.rs]: If we do take `entries` in as a `mut Vec`, we can avoid cloning here and just do:
```rust
let dup = entries
    .drain(..)
    .filter(|(key, _value)| existing.contains(&key))
    .collect();
```
and likewise at the end of the method, we could just do
```rust
self.inner.data.extend(entries);
```

File: src/errors.rs (hunk 3 above)
Comment (Fraser999) [reply src/errors.rs]: I guess to avoid confusion with crypto keys though, this could maybe be renamed to `DuplicateEntryKeys` rather than just `DuplicateKeys`?

File: src/append_only_data.rs [hunk 8]
```diff
@@ -658,15 +658,39 @@ impl UnseqAppendOnlyData<UnpubPermissions> {
...
+    Ok(())
+}
+
 impl<P> SeqAppendOnly for SeqAppendOnlyData<P>
 where
     P: Permissions + std::hash::Hash + Clone,
 {
-    fn append(&mut self, entries: &[(Vec<u8>, Vec<u8>)], last_entries_index: u64) -> Result<()> {
+    fn append(&mut self, mut entries: Entries, last_entries_index: u64) -> Result<()> {
+        match check_dup(&self.inner.data, entries.as_mut()) {
```
Comment (nbaksalyar) [inline src/append_only_data.rs]: Probably can be replaced with just
```suggestion
        if let Err(e) = check_dup(&self.inner.data, entries.as_mut()) {
            return Err(e);
        }
```

File: src/append_only_data.rs [hunk 9]
```diff
@@ -675,8 +699,13 @@ impl<P> UnseqAppendOnly for UnseqAppendOnlyData<P>
 where
     P: Permissions + std::hash::Hash + Clone,
 {
-    fn append(&mut self, entries: &[(Vec<u8>, Vec<u8>)]) -> Result<()> {
-        self.inner.data.extend(entries.iter().cloned());
+    fn append(&mut self, mut entries: Entries) -> Result<()> {
+        match check_dup(&self.inner.data, entries.as_mut()) {
```
Comment (nbaksalyar) [inline src/append_only_data.rs]: Same here, I think `if let` should suffice?

File: src/append_only_data.rs [hunk 10]
```diff
@@ -969,4 +998,15 @@ mod tests {
...
+
+    #[test]
+    fn append_unseq_test() {
+        let mut data = UnpubUnseqAppendOnlyData::new(XorName(rand::random()), 10);
+        let key1 = b"KEY1".to_vec();
+        let val1 = b"VALUE1".to_vec();
+        let key2 = b"KEY2".to_vec();
+        let val2 = b"VALUE2".to_vec();
+        let tup = [(key1, val1), (key2, val2)].to_vec();
+        assert_eq!(unwrap!(data.append(tup)), ());
```
Comment (nbaksalyar) [inline src/append_only_data.rs]: Can be
```suggestion
        assert!(data.append(tup).is_ok());
```

File: src/append_only_data.rs (hunk 8 above)
Comment (Fraser999) [reply src/append_only_data.rs]: Or even just
```rust
check_dup(&self.inner.data, entries.as_mut())?;
```

File: src/append_only_data.rs [hunk 11]
```diff
@@ -931,7 +960,7 @@ mod tests {
     #[test]
     fn seq_append_entries() {
         let mut data = SeqAppendOnlyData::<PubPermissions>::new(XorName([1; 32]), 10000);
-        let res = data.append(&[(b"hello".to_vec(), b"world".to_vec())], 0);
+        let res = data.append([(b"hello".to_vec(), b"world".to_vec())].to_vec(), 0);
```
Comment (Fraser999) [inline src/append_only_data.rs]: A minor point: this could be `vec![...]` rather than `[...].to_vec()`.

Also, we seem to have a few instances of
```rust
match res {
    Ok(()) => (),
    Err(x) => panic!("Unexpected error: {:?}", x),
}
```
These could all be removed if we just do e.g.
```rust
unwrap!(data.append(vec![(b"hello".to_vec(), b"world".to_vec())], 0));
```

File: src/append_only_data.rs (hunk 10 above)
Comment (Fraser999) [reply src/append_only_data.rs]: I agree the `assert_eq` in unnecessary, but I generally prefer unwrapping results in tests rather than just checking `is_ok()` since if it fails, you see the error value.  Not a big deal though.

File: src/errors.rs [hunk 12]
```diff
@@ -40,6 +41,10 @@ pub enum Error {
     InvalidEntryActions(BTreeMap<Vec<u8>, EntryError>),
     /// Key does not exist
     NoSuchKey,
+    /// Key already exists
+    KeysExists(Entries),
```
Comment (Fraser999) [inline src/errors.rs]: Should be `KeysExist`.  Also, the doc comment could probably be expanded a bit, e.g.
```rust
/// The key(s) of the entry or entries contained in this error already exist
```

File: src/append_only_data.rs [hunk 13]
```diff
@@ -658,15 +658,38 @@ impl UnseqAppendOnlyData<UnpubPermissions> {
...
+    Ok(())
+}
+
 impl<P> SeqAppendOnly for SeqAppendOnlyData<P>
 where
     P: Permissions + std::hash::Hash + Clone,
 {
-    fn append(&mut self, entries: &[(Vec<u8>, Vec<u8>)], last_entries_index: u64) -> Result<()> {
+    fn append(&mut self, mut entries: Entries, last_entries_index: u64) -> Result<()> {
+        if let Err(e) = check_dup(&self.inner.data, entries.as_mut()) {
```
Comment (Fraser999) [inline src/append_only_data.rs]: This can be just:
```rust
check_dup(&self.inner.data, entries.as_mut())?;
```
Applies in another place below too.

File: src/append_only_data.rs [hunk 14]
```diff
@@ -969,4 +996,64 @@ mod tests {
...
+        let mut data = UnpubUnseqAppendOnlyData::new(XorName(rand::random()), 10);
+
+        // Assert that the current Append Operation has duplicate entries present and are not pushed
+        let entries = [
+            (b"KEY1".to_vec(), b"VALUE1".to_vec()),
+            (b"KEY2".to_vec(), b"VALUE2".to_vec()),
+            (b"KEY1".to_vec(), b"VALUE1".to_vec()),
+        ]
+        .to_vec();
+        assert!(data.append(entries).is_err());
```
Comment (Fraser999) [inline src/append_only_data.rs]: It'd be more thorough to check we're getting the correct variant returned:
```rust
assert_eq!(Error::DuplicateEntryKeys, unwrap_err!(data.append(entries)));
```
Also, it might be worth checking the state of `data.entries()` is as expected?  Even just once at the very end of the test, or after each attempt to mutate the data.

File: src/append_only_data.rs [hunk 15]
```diff
@@ -1026,4 +1045,70 @@ mod tests {
         let decoded = unwrap!(self::Address::decode_from_zbase32(&encoded));
         assert_eq!(address, decoded);
     }
+
+    #[test]
+    fn append_unseq_data_test() {
+        let mut data = UnpubUnseqAppendOnlyData::new(XorName(rand::random()), 10);
+
+        // Assert that the current Append Operation has duplicate entries present and are not pushed
```
Comment (nbaksalyar) [inline src/append_only_data.rs]: Maybe rephrase it a bit? Also, we're looking for duplicate _keys_ as opposed to entries (keys + values).
```suggestion
        // Assert that the entries are not appended because of duplicate keys.
```

File: src/append_only_data.rs [hunk 16]
```diff
@@ -1026,4 +1045,70 @@ mod tests {
...
+        let entries = vec![
+            (b"KEY1".to_vec(), b"VALUE1".to_vec()),
+            (b"KEY2".to_vec(), b"VALUE2".to_vec()),
+            (b"KEY1".to_vec(), b"VALUE1".to_vec()),
+        ];
+        assert_eq!(Error::DuplicateEntryKeys, unwrap_err!(data.append(entries)));
+
+        // Assert that the current Append Operation has no duplicate entries present and are pushed
```
Comment (nbaksalyar) [inline src/append_only_data.rs]: Maybe rephrase it a bit? Also, we're looking for duplicate keys as opposed to entries (keys + values).
```suggestion
        // Assert that the entries are appended because there are no duplicate keys.
```

File: src/append_only_data.rs [hunk 17]
```diff
@@ -1026,4 +1045,70 @@ mod tests {
...
+        let entries1 = vec![
+            (b"KEY1".to_vec(), b"VALUE1".to_vec()),
+            (b"KEY2".to_vec(), b"VALUE2".to_vec()),
+        ];
+
+        unwrap!(data.append(entries1));
+
+        // Assert that duplicate entries present and are not pushed
```
Comment (nbaksalyar) [inline src/append_only_data.rs]:
```suggestion
        // Assert that entries are not appended because they duplicate some keys appended previously.
```

File: src/append_only_data.rs [hunk 18]
```diff
@@ -1026,4 +1045,70 @@ mod tests {
...
+        let entries2 = vec![(b"KEY2".to_vec(), b"VALUE2".to_vec())];
+        assert_eq!(
+            Error::KeysExist(entries2.clone()),
+            unwrap_err!(data.append(entries2))
+        );
+
+        // Assert that duplicate entries not present and are pushed
```
Comment (nbaksalyar) [inline src/append_only_data.rs]:
```suggestion
        // Assert that no duplicate keys are present and the append operation is successful.
```

File: src/append_only_data.rs [hunk 19]
```diff
@@ -1026,4 +1045,70 @@ mod tests {
...
+        let entries3 = vec![(b"KEY3".to_vec(), b"VALUE3".to_vec())];
+        unwrap!(data.append(entries3));
+    }
+
+    #[test]
+    fn append_seq_data_test() {
+        let mut data = UnpubSeqAppendOnlyData::new(XorName(rand::random()), 10);
+
+        // Assert that the current Append Operation has duplicate entries present and are not pushed
```
Comment (nbaksalyar) [inline src/append_only_data.rs]:
```suggestion
        // Assert that the entries are not appended because of duplicate keys.
```

File: src/append_only_data.rs [hunk 20]
```diff
@@ -1026,4 +1045,70 @@ mod tests {
...
+            (b"KEY1".to_vec(), b"VALUE1".to_vec()),
+            (b"KEY2".to_vec(), b"VALUE2".to_vec()),
+            (b"KEY1".to_vec(), b"VALUE1".to_vec()),
+        ];
+        assert_eq!(
+            Error::DuplicateEntryKeys,
+            unwrap_err!(data.append(entries, 0))
+        );
+
+        // Assert that the current Append Operation has no duplicate entries present and are pushed
```
Comment (nbaksalyar) [inline src/append_only_data.rs]:
```suggestion
        // Assert that the entries are appended because there are no duplicate keys.
```

File: src/append_only_data.rs [hunk 21]
```diff
@@ -1026,4 +1045,70 @@ mod tests {
...
+        let entries1 = vec![
+            (b"KEY1".to_vec(), b"VALUE1".to_vec()),
+            (b"KEY2".to_vec(), b"VALUE2".to_vec()),
+        ];
+        unwrap!(data.append(entries1, 0));
+
+        // Assert that duplicate entries present and are not pushed
```
Comment (nbaksalyar) [inline src/append_only_data.rs]:
```suggestion
        // Assert that entries are not appended because they duplicate some keys appended previously.
```

File: src/append_only_data.rs [hunk 22]
```diff
@@ -1026,4 +1045,70 @@ mod tests {
...
+        let entries2 = vec![(b"KEY2".to_vec(), b"VALUE2".to_vec())];
+        assert_eq!(
+            Error::KeysExist(entries2.clone()),
+            unwrap_err!(data.append(entries2, 2))
+        );
+
+        // Assert that duplicate entries not present and are pushed
```
Comment (nbaksalyar) [inline src/append_only_data.rs]:
```suggestion
        // Assert that no duplicate keys are present and the append operation is successful.
```
